server/aws-lsp-codewhisperer/src/language-server/agenticChat/errors.ts

```diff
@@ -71,3 +71,194 @@ export function isRequestAbortedError(error: unknown): boolean {
+        )
+    }
+}
+
+export class TextNotFoundError extends FileOperationError {
+    constructor(text: string) {
+        super(
+            `No occurrences of "${text}" were found`,
+            'The text to replace was not found in the file. Please check your input.'
+        )
+    }
+}
+
+export class MultipleMatchesError extends FileOperationError {
+    constructor(text: string) {
+        super(
+            `Multiple occurrences of "${text}" were found when only 1 is expected`,
+            'Multiple instances of the text to replace were found. Please provide more context to identify the specific text to replace.'
+        )
+    }
+}
+
+/**
+ * Maps a standard Error to the appropriate FileOperationError type
+ * @param error The original error
+ * @returns A FileOperationError with customer-facing message
+ */
+export function createFileOperationError(error: Error): FileOperationError {
+    const message = error.message
+
+    // Common file system errors
+    if (message.includes('ENOENT') || message.includes('no such file or directory')) {
+        return new DirectoryNotFoundError(message)
+    }
+    if (message.includes('EACCES') || message.includes('permission denied')) {
+        return new PermissionDeniedError(message)
+    }
+    if (message.includes('EISDIR')) {
+        return new IsDirectoryError(message)
+    }
+    if (message.includes('ENOSPC')) {
+        return new NoSpaceError(message)
+    }
+    if (message.includes('EMFILE') || message.includes('ENFILE')) {
+        return new TooManyOpenFilesError(message)
+    }
+
+    // fsWrite specific errors
+    if (message === 'Path must not be empty') {
+        return new EmptyPathError()
+    }
+    if (message === 'fileText must be provided for create command') {
+        return new MissingContentError()
+    }
+    if (message === 'The file already exists with the same content') {
+        return new FileExistsWithSameContentError()
+    }
+    if (message === 'Content to append must not be empty') {
+        return new EmptyAppendContentError()
+    }
+
+    // fsReplace specific errors
+    if (message === 'Diffs must not be empty') {
+        return new EmptyDiffsError()
+    }
+    if (message === 'The provided path must exist in order to replace contents into it') {
+        return new FileNotExistsError()
+    }
+
+    // Pattern matching for errors with dynamic content
+    const noOccurrencesMatch = message.match(/No occurrences of "(.+)" were found/)
+    if (noOccurrencesMatch) {
+        return new TextNotFoundError(noOccurrencesMatch[1])
+    }
+
+    const multipleOccurrencesMatch = message.match(/Multiple occurrences of "(.+)" were found/)
+    if (multipleOccurrencesMatch) {
+        return new MultipleMatchesError(multipleOccurrencesMatch[1])
+    }
+
+    // Default fallback - wrap the original error with the same message
+    return new FileOperationError(message, message)
```

Comment:
what about invalid json?

Reply:
The invalid json case won't happen here, that will be caught outside of the tool altogether